File: php/src/Simple/Composability/LotteryOrder.php
<?php

namespace ImmutableDatamodel\Simple\Composability;

use ImmutableDatamodel\Shared\OrderConstraint;
use ImmutableDatamodel\Simple\Order;

readonly class LotteryOrder implements Order, Winnable
{
    /**
     * @param string $orderId
     * @param OrderConstraint[] $constraints
     */
    public function __construct(
        private string $orderId,
        private array  $constraints
    ) {
    }

    public function orderId(): string
    {
        return $this->orderId;
    }

    public function constraints(): array
    {
        return $this->constraints;
    }

    public function win(string $productSerialNo): WinningOrder
    {
        return new WinningOrder($this->orderId(), $this->constraints(), $productSerialNo);
    }
}

File: golang/src/simple/lottery_order.go
package simple

import "example/shared"

type LotteryOrder struct {
	orderId     string
	constraints []shared.OrderConstraint
}

func (lo LotteryOrder) OrderId() string {
	return lo.orderId
}

func (lo LotteryOrder) Constraints() []shared.OrderConstraint {
	return lo.constraints
}

func (lo LotteryOrder) Win(productSerialNo string) WinningOrder {
	return WinningOrder{orderId: lo.OrderId(), constraints: lo.Constraints(), productSerialNo: productSerialNo}
}
